### ISSUE TITLE
Deadlock on Shutdown

### ISSUE THREAD

**Issue by WernerderChamp** [iotaledger/hornet#910]
**Describe the bug**
I restarted hornet a short time after starting it up and it deadlocked on shutdown.
The node was pruning at the moment I rebooted it.
It might be related to #899 but the stack dump seems quite different

**Environment information:**
 - OS: Ubuntu 18.04
 - RAM: 4GB
 - Cores: 3 Cores
 - Type: VPS (Hetzner CPX31)
 - HORNET version: latest commit on develop (c951639eec2f302f8a6e5a3c9b756ba915ba76f7)

**Additional context**

[goroutine.txt](https://github.com/gohornet/hornet/files/6055144/goroutine.txt)
[logs.txt](https://github.com/gohornet/hornet/files/6055145/logs.txt)

**Comment by muXxer** [comment]
This commit hash can't be correct. Can you please check again.

**Comment by muXxer** [comment]
Ah I see, the flush lock in the object storage deadlocks with the Shutdown of the timed executor.... :rekt:
```
goroutine 197 [semacquire, 4 minutes]:
sync.runtime_Semacquire(0xc00c557418)
	/usr/local/go/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc00c557410)
	/usr/local/go/src/sync/waitgroup.go:130 +0x65
github.com/iotaledger/hive.go/timedexecutor.(*TimedExecutor).Shutdown(0xc00c557400, 0xc00c554e07, 0x1, 0x1)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/timedexecutor/timedexecutor.go:62 +0x8e
github.com/iotaledger/hive.go/objectstorage.(*ObjectStorage).flush(0xc007c510e0, 0xc000a91b01)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/objectstorage/object_storage.go:929 +0x78
github.com/iotaledger/hive.go/objectstorage.(*ObjectStorage).Shutdown(0xc007c510e0)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/objectstorage/object_storage.go:513 +0x38
github.com/gohornet/hornet/pkg/protocol/gossip.(*MessageProcessor).Run(0xc00c525260, 0xc00eed9bc0)
	/opt/hornet/pkg/protocol/gossip/msg_proc.go:133 +0x6a
github.com/gohornet/hornet/core/gossip.run.func5(0xc00eed9bc0)
	/opt/hornet/core/gossip/core.go:244 +0xa5
github.com/iotaledger/hive.go/daemon.(*OrderedDaemon).runBackgroundWorker.func1(0xc00003e380, 0x19d1ca6, 0x10, 0x1f55b30, 0xc00f049fe0, 0xc00f06b3a0)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/daemon/daemon.go:140 +0x53
created by github.com/iotaledger/hive.go/daemon.(*OrderedDaemon).runBackgroundWorker
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/daemon/daemon.go:136 +0x105

goroutine 159 [semacquire, 4 minutes]:
sync.runtime_SemacquireMutex(0xc007c51134, 0x0, 0x0)
	/usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*RWMutex).RLock(...)
	/usr/local/go/src/sync/rwmutex.go:50
github.com/iotaledger/hive.go/objectstorage.(*CachedObjectImpl).BatchWriteDone(0xc0076ba280)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/objectstorage/cached_object.go:373 +0x2ca
github.com/iotaledger/hive.go/objectstorage.(*CachedObjectImpl).evict(0xc0076ba280)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/objectstorage/cached_object.go:299 +0x73
github.com/iotaledger/hive.go/objectstorage.(*CachedObjectImpl).delayedRelease(0xc0076ba280)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/objectstorage/cached_object.go:123 +0x54
github.com/iotaledger/hive.go/timedexecutor.(*TimedExecutor).startBackgroundWorkers.func1(0xc00c557400)
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/timedexecutor/timedexecutor.go:71 +0x4c
created by github.com/iotaledger/hive.go/timedexecutor.(*TimedExecutor).startBackgroundWorkers
	/root/go/pkg/mod/github.com/muxxer/hive.go@v0.0.0-20210222004711-d924b9529a49/timedexecutor/timedexecutor.go:69 +0x5f
```

**Comment by WernerderChamp** [comment]
> This commit hash can't be correct. Can you please check again.

Oops, I fixed it. I was on the wrong branch when looking for the hash 🤦

**Comment by muXxer** [comment]
Fixed with #919
